### PR TITLE
gnrc: 6lowpan-nd: some broken router dependencies

### PR DESCRIPTION
**Pull request by OlegHahm**
Please check thoroughly if I got that right.

### CODE REVIEW

File: sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c [hunk 1]
```diff
@@ -243,8 +243,9 @@ void gnrc_ipv6_netif_set_router(gnrc_ipv6_netif_t *netif, bool enable)
         gnrc_sixlowpan_nd_router_set_router(netif, enable);
         return;
     }
-#endif
+#else
```
Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: Particular here I wasn't sure if this was on purpose. But I thought pulling in `gnrc_ndp_router` seems odd if it's only needed for this one function.

Comment (authmillenon) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: Nope if both modules are present it should call both functions.

File: sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c [hunk 2]
```diff
@@ -776,6 +777,10 @@ void gnrc_ipv6_netif_init_by_dev(void)
 
             DEBUG("ipv6 netif: Set 6LoWPAN flag\n");
             ipv6_ifs[i].flags |= GNRC_IPV6_NETIF_FLAGS_SIXLOWPAN;
+#ifdef MODULE_GNRC_SIXLOWPAN_ND_ROUTER
+            DEBUG("ipv6 netif: Set router flag\n");
+            ipv6_ifs[i].flags |= GNRC_IPV6_NETIF_FLAGS_ROUTER;
```
Comment (authmillenon) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: Will be set (with correct send-out behavior) in https://github.com/RIOT-OS/RIOT/pull/3813/files#diff-f9bf85cb337616822921e4230ede1572R830

File: sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c (hunk 1 above)
Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: So, the correct solution would be another ifdef here? Currently gnrc_sixlowpan_nd_router won't link.

File: sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c (hunk 2 above)
Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: Somehow the link seems to be broken (or at least I cannot see).

File: sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c (hunk 1 above)
Comment (authmillenon) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: Yes.

File: sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c (hunk 2 above)
Comment (authmillenon) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: `gnrc_ipv6_netif_set_router()` is called in L830 which does the same thing (+ other stuff) already.

Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: But that's "too late" to set the solicited node address.

File: sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c (hunk 1 above)
Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: Updated in the latest commit, but can you tell me in which case a node would have both modules enabled? Border router?

File: sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c (hunk 2 above)
Comment (authmillenon) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: Ah, then it needs to be moved.

Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: Can I call `gnrc_ipv6_netif_set_router()` before the node has an address configured?

Comment (authmillenon) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: No.

Comment (authmillenon) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: Please comment on that.

Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: done

File: sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c [hunk 3]
```diff
@@ -776,6 +778,16 @@ void gnrc_ipv6_netif_init_by_dev(void)
 
             DEBUG("ipv6 netif: Set 6LoWPAN flag\n");
             ipv6_ifs[i].flags |= GNRC_IPV6_NETIF_FLAGS_SIXLOWPAN;
+
+            /* the router flag must be set early here, because otherwise
+             * _add_addr_to_entry() wouldn't set the solicited node address.
+             * However, addresses has to be configured before calling
```
Comment (authmillenon) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: s/has/have/

File: sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c [hunk 4]
```diff
@@ -238,13 +238,15 @@ gnrc_ipv6_netif_t *gnrc_ipv6_netif_get(kernel_pid_t pid)
...
 void gnrc_ipv6_netif_set_router(gnrc_ipv6_netif_t *netif, bool enable)
 {
-#if defined(MODULE_GNRC_SIXLOWPAN_ND_ROUTER)
+#ifdef MODULE_GNRC_SIXLOWPAN_ND_ROUTER
     if (netif->flags & GNRC_IPV6_NETIF_FLAGS_SIXLOWPAN) {
         gnrc_sixlowpan_nd_router_set_router(netif, enable);
         return;
     }
 #endif
+#ifdef MODULE_GNRC_NDP_ROUTER
```
Comment (authmillenon) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: L260 needs the same treatment

**Comment by authmillenon** [comment]
I get ping timeouts with this PR, even if I set the rtr_adv flag manually.

**Comment by OlegHahm** [comment]
Can you specify your setting? Cannot reproduce the problem.

**Comment by authmillenon** [comment]
I added `gnrc_ipv6_netif_set_rtr_adv()` after `gnrc_ipv6_netif_set_router()` in `gnrc_ipv6_netif_init_by_dev()` and compiled `gnrc_networking` as normal, otherwise. I only did that because it wasn't working with vanilla `gnrc_networking` either.

**Comment by OlegHahm** [comment]
Which platform? It seems to be working on native but not iotlab-m3.

**Comment by OlegHahm** [comment]
Apparently no neighbor solicitations are sent.

**Comment by authmillenon** [comment]
On iotlab-m3. Will test tomorrow again

**Comment by OlegHahm** [comment]
`gnrc_sixlowpan_nd_wakeup()` is never called - it's not even linked in...

**Comment by authmillenon** [comment]
It's supposed to be called when the node wakes up. Since we don't have a sleep-cycle yet, it is of course not called yet.

File: sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c (hunk 4 above)
Comment (OlegHahm) [inline sys/net/gnrc/network_layer/ipv6/netif/gnrc_ipv6_netif.c]: addressed

**Comment by OlegHahm** [comment]
With #3828 pinging between two iotlab-m3 is working.

**Comment by authmillenon** [comment]
Yes it's working. Please squash.

**Comment by OlegHahm** [comment]
squashed

**Comment by authmillenon** [comment]
ACK

**Pull request merged**